File: packages/oxlint-plugin-react-doctor/src/plugin/rules/security-scan/clickjacking-redirect-risk.ts
import { defineRule } from "../../utils/define-rule.js";
import { isConfigOrCiPath } from "./utils/is-config-or-ci-path.js";
import { isProductionSourcePath } from "./utils/is-production-source-path.js";
import { scanByPattern } from "./utils/scan-by-pattern.js";

export const clickjackingRedirectRisk = defineRule({
  id: "clickjacking-redirect-risk",
  title: "Redirect or frame boundary risk",
  severity: "warn",
  recommendation:
    "Allowlist redirect origins/paths, set `frame-ancestors` for privileged pages, and avoid URL-prefilled privileged dialogs.",
  // `[^)'"\n]*` keeps redirect keywords inside string literals
  // (`redirect('/login?message=...continue...')`) from counting — only a bare
  // identifier/property between `redirect(` and the keyword is caller input.
  // `(?!\s*(?:await\s+)?[\w$]*(?:safe|valid|sanitiz|allowlist|whitelist)`
  // skips targets already passed through a safe-redirect helper.
  // The iframe branch requires URL-param shapes (`redirect=`, query-style
  // `role=` — `[?&]role=`, including the `&amp;`-escaped form), not the bare
  // word `redirect` or an ARIA `role="..."` attribute — the 700-char window
  // bleeds past the tag, so a bare `role=` matched the accessible-iframe /
  // iframe-inside-a-modal idioms.
  scan: scanByPattern({
    shouldScan: (file) =>
      isProductionSourcePath(file.relativePath) || isConfigOrCiPath(file.relativePath),
    pattern:
      /\bredirect\s*\((?!\s*(?:await\s+)?[\w$]*(?:safe|valid|sanitiz|allowlist|whitelist)[\w$]*\s*\()[^)'"`\n]*\b(?:searchParams\.get|nextUrl\.searchParams|returnTo|callbackUrl|continue|next)\b|<iframe\b[\s\S]{0,700}(?:\b(?:next=|continue=|redirect=|redirect_uri|userstoinvite|sharingaction|\.\.)|[?&](?:amp;)?role=)|frame-ancestors\s+(?:\*|'self'\s+\*)|X-Frame-Options["']?\s*:\s*["']?ALLOW/i,
    message:
      "Redirect or framing configuration may let attacker-controlled URLs chain into privileged UI or clickjacking.",
  }),
});

File: native/oxlint/scans/clickjacking-redirect-risk.rs
use lazy_regex::{Lazy, Regex, lazy_regex};

use super::{ScanFinding, get_location_at_index::get_location_at_index};

const MESSAGE: &str = "Redirect or framing configuration may let attacker-controlled URLs chain into privileged UI or clickjacking.";

static REDIRECT_CALL_PATTERN: Lazy<Regex> = lazy_regex!(r"(?i)\bredirect\s*\(");
static SAFE_REDIRECT_PATTERN: Lazy<Regex> = lazy_regex!(
    r"(?i)^\s*(?:await\s+)?[A-Za-z0-9_$]*(?:safe|valid|sanitiz|allowlist|whitelist)[A-Za-z0-9_$]*\s*\("
);
static REDIRECT_INPUT_PATTERN: Lazy<Regex> = lazy_regex!(
    r"(?i)\b(?:searchParams\.get|nextUrl\.searchParams|returnTo|callbackUrl|continue|next)\b"
);
static IFRAME_PATTERN: Lazy<Regex> = lazy_regex!(
    r"(?is)<iframe\b.{0,700}(?:\b(?:next=|continue=|redirect=|redirect_uri|userstoinvite|sharingaction|\.\.)|[?&](?:amp;)?role=)"
);
static FRAME_POLICY_PATTERN: Lazy<Regex> = lazy_regex!(
    r#"(?i)frame-ancestors\s+(?:\*|'self'\s+\*)|X-Frame-Options["']?\s*:\s*["']?ALLOW"#
);

pub fn scan(relative_path: &str, source: &str) -> Vec<ScanFinding> {
    if !super::is_production_file_path::is_production_source_path(relative_path)
        && !super::is_config_or_ci_path::is_config_or_ci_path(relative_path)
    {
        return Vec::new();
    }
    let content = super::get_scannable_content::get_scannable_content(relative_path, source, false);
    let normalized = super::normalize_js_regex_content::normalize_js_regex_content(&content);

    let redirect_index = REDIRECT_CALL_PATTERN
        .find_iter(&normalized)
        .find_map(|redirect_match| {
            let arguments_start = redirect_match.end();
            let tail = &normalized[arguments_start..];
            if SAFE_REDIRECT_PATTERN.is_match(tail) {
                return None;
            }
            let argument_end = tail
                .char_indices()
                .find(|(_, character)| matches!(character, ')' | '\'' | '"' | '`' | '\n'))
                .map_or(tail.len(), |(index, _)| index);
            if REDIRECT_INPUT_PATTERN.is_match(&tail[..argument_end]) {
                return Some(redirect_match.start());
            }
            None
        });
    let other_index = IFRAME_PATTERN
        .find(&normalized)
        .map(|found| found.start())
        .into_iter()
        .chain(
            FRAME_POLICY_PATTERN
                .find(&normalized)
                .map(|found| found.start()),
        )
        .min();
    redirect_index
        .into_iter()
        .chain(other_index)
        .min()
        .map(|index| finding(source, &content, index))
        .unwrap_or_default()
}

fn finding(original: &str, content: &str, index: usize) -> Vec<ScanFinding> {
    let (line, column) = get_location_at_index(original, content, index);
    vec![ScanFinding::inherited(MESSAGE, line, column)]
}
